### PR TITLE
🌐 Add MVP for translation in the Supervisor panel

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

- Adds `.localize` to the `Supervisor` object and all the things around that to make it work.
- Migrates all the tabs/dashboards inside the panel to use `this.supervisor.localize(...)`
- The content on each dashboard is not migrated in this to keep it as small as possible, with the exception of the main dashboard.
- Changes the layout of the update cards to be more like the cards under the system tab.
- Names will not be added to translations;
   - Home Assistant
   - Home Assistant Core
   - Core
   - Supervisor
   - Home Assistant Supervisor
   - Operating System
   - Home Assistant Operating System

Once this is added, folks that want to help with translation of the Supervisor panel can join the frontend team on localize <https://developers.home-assistant.io/docs/translations>, new keys will be available in localize when more of the Supervisor panel is migrated.

**Feature requests**: 
- <https://community.home-assistant.io/t/home-assistant-supervisor-translation/186371>
- <https://community.home-assistant.io/t/hope-to-start-supervisor-localization/180129>
- <https://community.home-assistant.io/t/lokalise-home-assistant-supervisor/185194>

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

unknown
image

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/translations/en.json [hunk 1]
```diff
@@ -3367,5 +3367,47 @@
...
+      }
+    },
+    "dashboard": {
+      "addon_new_version": "New version available",
+      "addon_running": "Add-on is running",
+      "addon_stopped": "Add-on is stopped",
+      "addons": "Add-ons",
+      "no_addons": "You don't have any add-ons installed yet. Head over to the add-on store to get started!",
+      "update_available": "Update pending {icon}",
+      "updates_available": "Updates pending {icon}"
```
Comment (bramkragten) [inline src/translations/en.json:3400]:
```suggestion
      "update_available": "{count, plural,\n  one {Update}\n  other {{count} Updates}\n} pending {icon}",
```

File: src/util/common-translation.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+  // Final fallback
+  return "en";
+}
+
+// Store loaded translations in memory so translations are available immediately
+// when DOM is created in Polymer. Even a cache lookup creates noticeable latency.
+const translations = {};
+
+async function fetchTranslation(fingerprint: string, supervisor: boolean) {
```
Comment (bramkragten) [inline src/util/common-translation.ts:100]: Everything above this function is just used in Home Assistant right? So we should put in in different file (like `hass-translation.ts`) so it is not imported for the supervisor.

File: src/translations/en.json (hunk 1 above)
Comment (bramkragten) [reply src/translations/en.json:3400]: You can use plural syntax in translations, as not all languages have just one/other :-)

Comment (ludeeus) [reply src/translations/en.json:3400]: That looks like magic

Comment (bramkragten) [reply src/translations/en.json:3400]: <https://formatjs.io/docs/intl-messageformat#common-usage-example>

File: hassio/src/dashboard/hassio-update.ts [hunk 3]
```diff
@@ -59,8 +68,16 @@ export class HassioUpdate extends LitElement {
...
+            ? this.supervisor.localize(
+                "dashboard.updates_available",
+                "icon",
+                "🎉"
+              )
+            : this.supervisor.localize(
+                "dashboard.update_available",
+                "icon",
+                "🎉"
+              )}
```
Comment (bramkragten) [inline hassio/src/dashboard/hassio-update.ts:80]:
```ts
            this.supervisor.localize(
                "dashboard.updates_available",
				"count",
				updatesAvailable,
                "icon",
                "🎉"
              )
```

File: src/translations/en.json [hunk 4]
```diff
@@ -3396,8 +3396,7 @@
       "addon_stopped": "Add-on is stopped",
       "addons": "Add-ons",
       "no_addons": "You don't have any add-ons installed yet. Head over to the add-on store to get started!",
-      "update_available": "Update pending {icon}",
-      "updates_available": "Updates pending {icon}"
+      "update_available": "{count, plural,\n  one {Update}\n  other {{count} Updates}\n} pending {icon}",
```
Comment (bramkragten) [inline src/translations/en.json:3399]: I think my {count} should be #... If you want to display the number of updated that is...
```
      "update_available": "{count, plural,\n  one {Update}\n  other {# Updates}\n} pending {icon}",

```

File: hassio/src/supervisor-base-element.ts [hunk 5]
```diff
@@ -61,9 +72,31 @@ export class SupervisorBaseElement extends urlSyncMixin(
 
   protected firstUpdated(changedProps: PropertyValues): void {
     super.firstUpdated(changedProps);
+    this._language = this.hass.language;
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:75]: We should check in updated for when this changes

File: hassio/src/supervisor-base-element.ts [hunk 6]
```diff
@@ -61,9 +72,31 @@ export class SupervisorBaseElement extends urlSyncMixin(
...
   }
 
+  private async _initLocalize() {
+    const { language, data } = await getTranslation(
+      "supervisor",
+      this._language,
+      true
+    );
+
+    this._localize = await computeLocalize(
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:86]: Why do we store this in the class?

File: hassio/src/supervisor-base-element.ts [hunk 7]
```diff
@@ -61,9 +72,31 @@ export class SupervisorBaseElement extends urlSyncMixin(
...
+      this.constructor.prototype,
+      this._language,
+      {
+        [language]: data,
+      }
+    );
+
+    this.supervisor = {
+      ...this.supervisor!,
+      localize: (key: string, ...args: any[]) => this._localize(key, ...args),
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:96]: This would also work right?
```ts
      localize: await computeLocalize(...,
```

Comment (ludeeus) [reply hassio/src/supervisor-base-element.ts:96]: If it does, that would be cleaner

File: hassio/src/supervisor-base-element.ts [hunk 8]
```diff
@@ -36,18 +41,24 @@ declare global {
...
 }
 
+const empty = () => "";
+
 export class SupervisorBaseElement extends urlSyncMixin(
   ProvideHassLitMixin(LitElement)
 ) {
   @property({ attribute: false }) public supervisor?: Supervisor;
 
+  @internalProperty() private _localize: LocalizeFunc = empty;
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:51]:
```suggestion
  @internalProperty() private _localize: LocalizeFunc = () => "";
```

Comment (ludeeus) [reply hassio/src/supervisor-base-element.ts:51]: this is being removed?, so change not needed

File: hassio/src/supervisor-base-element.ts [hunk 9]
```diff
@@ -110,42 +143,44 @@ export class SupervisorBaseElement extends urlSyncMixin(
...
-    };
+      this.addEventListener("supervisor-update", (ev) =>
+        this._updateSupervisor(ev.detail)
+      );
+    }
 
-    this.addEventListener("supervisor-update", (ev) =>
-      this._updateSupervisor(ev.detail)
-    );
+    await this._initLocalize();
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:184]: We should start this sooner, and we don't have to await it?

File: src/util/common-translation.ts (hunk 2 above)
Comment (bramkragten) [inline src/util/common-translation.ts:100]:
```suggestion
async function fetchTranslation(fingerprint: string, supervisor?: boolean) {
```

File: src/util/common-translation.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  const response = await fetch(
+    supervisor
+      ? `/api/hassio/app/static/translations/${fingerprint.replace(
+          "supervisor/",
+          ""
+        )}`
```
Comment (bramkragten) [inline src/util/common-translation.ts:106]: Shouldnt this be the task of the supervisor?

Comment (bramkragten) [reply src/util/common-translation.ts:106]: Maybe also replace the `supervisor` boolean with an optional `basePath`?

File: hassio/src/supervisor-base-element.ts (hunk 8 above)
Comment (bramkragten) [reply hassio/src/supervisor-base-element.ts:51]: we would still need to init the supervisor with a stub localize function, for when we render before the translations are loaded. We didn't do that before btw :-)

File: src/util/common-translation.ts (hunk 10 above)
Comment (ludeeus) [reply src/util/common-translation.ts:106]: The replacement will be removed, I will adjust the gulp task to do it there.
Sure, can do basepath

File: hassio/src/supervisor-base-element.ts [hunk 11]
```diff
@@ -49,20 +51,71 @@ export class SupervisorBaseElement extends urlSyncMixin(
...
+      if (changedProperties.get("_language") !== this._language) {
+        this._initializeLocalize();
+      }
+    }
+
+    if (
+      this._language &&
+      this._resources &&
+      (changedProperties.has("_language") ||
+        changedProperties.has("_resources"))
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:83]: This means that we will `computeLocalize` twice when the language changes, as the `_initializeLocalize` from line 75 will also create an update when it has loaded the resources?

File: hassio/src/supervisor-base-element.ts [hunk 12]
```diff
@@ -40,7 +42,7 @@ declare global {
 export class SupervisorBaseElement extends urlSyncMixin(
   ProvideHassLitMixin(LitElement)
 ) {
-  @property({ attribute: false }) public supervisor?: Supervisor;
+  @property({ attribute: false }) public supervisor: Partial<Supervisor> = {};
```
Comment (bramkragten) [inline hassio/src/supervisor-base-element.ts:45]:
```suggestion
  @property({ attribute: false }) public supervisor: Partial<Supervisor> = {localize: () => ""};
```

File: hassio/src/hassio-main.ts [hunk 13]
```diff
@@ -72,7 +77,7 @@ export class HassioMain extends SupervisorBaseElement {
   }
 
   protected render() {
-    if (!this.supervisor || !this.hass) {
+    if (!this.supervisor.localize || !this.hass) {
```
Comment (bramkragten) [inline hassio/src/hassio-main.ts:80]: Why do we wait for localize? We should just start rendering?

Comment (ludeeus) [reply hassio/src/hassio-main.ts:80]: On reconnect, translations are not properly loaded (if you switch from SU to configuration and back)

Comment (bramkragten) [reply hassio/src/hassio-main.ts]: Don't we update when the translations are loaded?

Comment (ludeeus) [reply hassio/src/hassio-main.ts]: Yes, but no, there is some weird stuff happening, this small adjustment makes it work 100% of the times

Comment (ludeeus) [reply hassio/src/hassio-main.ts]: fund the issue, will adjust and update

**Review by bramkragten — APPROVED** [review]

**Pull request merged**